base/abstractarray.jl

```diff
@@ -2702,134 +2702,152 @@ foreach(f, itrs...) = (for z in zip(itrs...); f(z...); end; nothing)
- 10  12
+[:, :, 3] =
+ 21  23  25  27  29
+ 22  24  26  28  30
 
-[:, :, 2, 2] =
- 13  15
- 14  16
+julia> f(x::Matrix) = fill(x[1,1], 1,4);  # returns a 1×4 matrix
 
-julia> mapslices(sum, a, dims = [1,2])
-1×1×2×2 Array{Int64, 4}:
-[:, :, 1, 1] =
- 10
+julia> mapslices(f, A, dims=(1,2))
+1×4×3 Array{$Int, 3}:
+[:, :, 1] =
+ 1  1  1  1
 
-[:, :, 2, 1] =
- 26
+[:, :, 2] =
+ 11  11  11  11
+
+[:, :, 3] =
+ 21  21  21  21
+
+julia> g(x) = x[begin] // x[end-1];  # returns a number
+
+julia> mapslices(g, A, dims=[1,3])
+1×5×1 Array{Rational{$Int}, 3}:
+[:, :, 1] =
+ 1//21  3//23  1//5  7//27  9//29
 
-[:, :, 1, 2] =
- 42
+julia> map(g, eachslice(A, dims=2))
+5-element Vector{Rational{$Int}}:
+ 1//21
+ 3//23
+ 1//5
+ 7//27
+ 9//29
 
-[:, :, 2, 2] =
- 58
+julia> mapslices(sum, A; dims=(1,3)) == sum(A; dims=(1,3))
+true
 ```
 """
 function mapslices(f, A::AbstractArray; dims)
-    if isempty(dims)
-        return map(f,A)
-    end
-    if !isa(dims, AbstractVector)
-        dims = [dims...]
-    end
+    isempty(dims) && return map(f, A)
 
-    dimsA = [axes(A)...]
-    ndimsA = ndims(A)
-    alldims = [1:ndimsA;]
-
-    otherdims = setdiff(alldims, dims)
-
-    idx = Any[first(ind) for ind in axes(A)]
-    itershape   = tuple(dimsA[otherdims]...)
     for d in dims
-        idx[d] = Slice(axes(A, d))
+        d isa Integer || throw(ArgumentError("dimension must be integers, got $d"))
+        d >= 1 || throw(ArgumentError("dimension must be ≥ 1, got $d"))
+        # Indexing a matrix M[:,1,:] produces a 1-column matrix, but dims=(1,3) here
+        # would otherwise ignore 3, and slice M[:,i]. Previously this gave error:
+        # BoundsError: attempt to access 2-element Vector{Any} at index [3]
+        d > ndims(A) && throw(ArgumentError("mapslices does not accept dims > ndims(A)"))
     end
+    dim_mask = ntuple(d -> d in dims, ndims(A))
 
     # Apply the function to the first slice in order to determine the next steps
-    Aslice = A[idx...]
+    idx1 = ntuple(d -> d in dims ? (:) : firstindex(A,d), ndims(A))
+    Aslice = A[idx1...]
     r1 = f(Aslice)
-    # In some cases, we can re-use the first slice for a dramatic performance
-    # increase. The slice itself must be mutable and the result cannot contain
-    # any mutable containers. The following errs on the side of being overly
-    # strict (#18570 & #21123).
-    safe_for_reuse = isa(Aslice, StridedArray) &&
-                     (isa(r1, Number) || (isa(r1, AbstractArray) && eltype(r1) <: Number))
 
-    # determine result size and allocate
-    Rsize = copy(dimsA)
-    # TODO: maybe support removing dimensions
-    if !isa(r1, AbstractArray) || ndims(r1) == 0
+    if r1 isa AbstractArray && ndims(r1) > 0
+        n = sum(dim_mask)
+        if ndims(r1) > n && any(ntuple(d -> size(r1,d+n)>1, ndims(r1)-n))
+            s = size(r1)[1:n]
+            throw(DimensionMismatch("cannot assign slice f(x) of size $(size(r1)) into output of size $s"))
+        end
+        res1 = r1
+    else
```

Comment:
Changing this loop to `res1 = if ... else ... end` seems to make the function type-stable.

However, it also produces a strange inference complaint:
```julia
julia> @inferred mapslices(cumsum, [1 2; 3 4], dims=1)  # unrelated example, now works fine
2×2 Matrix{Int64}:
 1  2
 3  4

julia> mapslices(tuple, [1 2; 3 4], dims=1)  # desired result, but does not infer
1×2 Matrix{Tuple{Vector{Int64}}}:
 ([1, 3],)  ([2, 4],)

julia> @inferred mapslices(tuple, [1 2; 3 4], dims=1)
ERROR: return type Matrix{Tuple{Vector{Int64}}} does not match inferred return type Matrix
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] top-level scope
   @ REPL[121]:1

julia> mapslices(x -> tuple(x), [1 2; 3 4], dims=1)  # wtf?
ERROR: fatal error in type inference (type bound)
Stacktrace:
 [1] mapslices(f::var"#278#279", A::Matrix{Int64}; dims::Int64)
   @ Main ./REPL[107]:31
 [2] top-level scope
   @ REPL[122]:1
```

Reply:
(Same "fatal error in type inference" on master of 13 december, I rebased to re-run CI.)

Reply:
Tests which hit this are marked broken in 62068c2

Reply:
(Apparently imperfectly, though!)